File: LinkedList/LinkedList.cpp
#include <iostream>
using namespace std;

// Node
class Node {
public:
    int key;
    Node* next;

    Node(int k): key(k), next(nullptr) {};
};

// Linked List
class LinkedList {
public:
    Node* head;
    Node* Search(int);
    void Traverse();
    void Insert(int, int);
    void Delete(int);

    LinkedList(): head(nullptr) {};
};

Node* LinkedList::Search(int k) {
    Node* ptr = this->head;
    while (ptr != nullptr) {
        if (ptr->key == k) break;
        ptr = ptr->next;
    }
    return ptr;
}

void LinkedList::Traverse() {
    Node* ptr = this->head;
    while (ptr != nullptr) {
        cout << ptr->key << " ";
        ptr = ptr->next;
    }
    cout << endl;
}

void LinkedList::Insert(int k, int i) {
    Node* newNode = new Node(k);

    // If empty
    if (this->head == nullptr)
        this->head = newNode;
    
    else {
        Node* ptr = this->head;
        for (int t=0; t<i; t++) {
            if (ptr->next == nullptr) break;
            ptr = ptr->next;
        }

        // Insert node at the beginning
        if (ptr == this->head) {
            newNode->next = this->head;
            this->head = newNode;
        }

        // Insert node at the end
        else if (ptr->next == nullptr)
            ptr->next = newNode;
        
        // Insert node in the middle
        else {
            newNode->next = ptr->next;
            ptr->next = newNode;
        }
    }
}

void LinkedList::Delete(int k) {
    Node* ptr = this->head;

    // Find location
    while (true) {
        if (ptr == nullptr) break;
        if (ptr->key == k || ptr->next->key == k) break;
        ptr = ptr->next;
    }

    // Failed
    if (ptr == nullptr);

    // Delete node at the beginning
    else if (ptr == this->head && ptr->key == k)
        this->head = this->head->next;
    
    // Delete node at the end
    else if (ptr->next->next == nullptr && ptr->next->key == k)
        ptr->next = nullptr;
    
    // Delete node in the middle
    else
        ptr->next = ptr->next->next;
}

File: LinkedList/LinkedList.py
# Node
class Node:
    def __init__(self, k, p=None):
        self.key = k
        self.next = p

# Linked List
class LinkedList:
    def __init__(self):
        self.head = None

    def Search(self, k):
        p = self.head
        while p != None:
            if p.key == k:
                break
            p = p.next
        return p
    
    def Traverse(self):
        p = self.head
        while p != None:
            print(f"{p.key} ")
            p = p.next
        print()

    def Insert(self, k, i):
        newNode = Node(k)
        
        # If empty
        if self.head == None:
            self.head = newNode

        else:
            p = self.head
            for _ in range(i):
                if p.next == None:
                    break
                p = p.next

            # Insert node at the beginning
            if p == self.head:
                newNode.next = self.head
                self.head = newNode

            # Insert node at the end
            elif p.next == None:
                p.next = newNode

            # Insert node in the middle
            else:
                newNode.next = p.next
                p.next = newNode

    def Delete(self, k):
        p = self.head

        # Find location
        while True:
            if p == None:
                break
            if p.key == k or p.next.key == k:
                break
            p = p.next
        
        # Failed
        if p == None:
            return

        # Delete node at the beginning
        elif p == self.head and p.key == k:
            self.head = self.head.next
        
        # Delete node at the end
        elif p.next.next == None and p.next.key == k:
            p.next = None
        
        # Delete node in the middle
        else:
            p.next = p.next.next